src/Compilers/CSharp/Portable/Symbols/Synthesized/InjectedNonNullTypesAttributeSymbol.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// Makes the System.Runtime.CompilerServices.NonNullTypesAttribute available in every compilation.
+    /// </summary>
+    internal sealed class InjectedNonNullTypesAttributeSymbol : InjectedAttributeSymbol
+    {
+        private ImmutableArray<CSharpAttributeData> _lazyCustomAttributes;
+
+        private InjectedNonNullTypesAttributeSymbol(
+            AttributeDescription description,
+            NamespaceSymbol containingNamespace,
+            CSharpCompilation compilation,
+            Func<CSharpCompilation, NamedTypeSymbol, DiagnosticBag, ImmutableArray<MethodSymbol>> getConstructors)
+            : base(description, containingNamespace, compilation, getConstructors)
+        {
+        }
+
+        public static InjectedNonNullTypesAttributeSymbol Create(NamespaceSymbol containingNamespace)
+        {
+            return new InjectedNonNullTypesAttributeSymbol(AttributeDescription.NonNullTypesAttribute, containingNamespace, containingNamespace.DeclaringCompilation, makeConstructor);
+
+            ImmutableArray<MethodSymbol> makeConstructor(CSharpCompilation compilation, NamedTypeSymbol containingType, DiagnosticBag diagnostics)
+            {
+                var boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
+
+                Binder.ReportUseSiteDiagnostics(boolType, diagnostics, Location.None);
+
+                var boolWithAnnotations = TypeSymbolWithAnnotations.Create(boolType);
+                // PROTOTYPE constructor should save the parameter into a field (for users of reflection)
+                return ImmutableArray.Create<MethodSymbol>(
+                    new NonNullTypesAttributeConstructorSymbol(
+                        containingType,
+                        m => ImmutableArray.Create(SynthesizedParameterSymbol.Create(m, boolWithAnnotations, 0, ConstantValue.True, name: "flag"))));
+            }
+        }
+
+        internal override void AddDiagnostics(DiagnosticBag recipient)
+        {
+            // pull on the attributes to collect their diagnostics too
+            _ = GetAttributes();
+            recipient.AddRange(_diagnostics);
+        }
+
+        public override ImmutableArray<CSharpAttributeData> GetAttributes()
+        {
+            if (_lazyCustomAttributes.IsDefault)
+            {
+                ImmutableInterlocked.InterlockedInitialize(ref _lazyCustomAttributes, MakeAttributes());
```

Comment:
>ImmutableInterlocked.InterlockedInitialize(ref _lazyCustomAttributes, MakeAttributes()); [](start = 16, length = 88)

It looks like there is a chance of duplicating diagnostics when more than one thread is executing this line at the same time.